### PR TITLE
Entity Event: Animal Tame

### PR DESCRIPTION
**Pull request by lazynessmind**
## What was added:

* Event to check when an animal is tamed.

### Fabric:

* Mixin on `TamableAnimal::tame(Player)` for:
  * Wolf, Cat, Parrot;
* Mixin on `RunLikeCrazyGoal::tick()` for:
  * Horses, Llamas, Donkeys, Mules;
* Mixin on `Ocelot::mobInteract(Player, InteractionHand)` for:
  * Ocelot;

### Forge:

* Subscribed to `AnimalTameEvent`

## Testing:

Since Forge isn't on 1.17 yet, I tested this implement on architectury-api 1.16 and it works. On Fabric 1.17 is all working fine.

`Tip:` If testing the Ocelot tame, you need to do it on survival mode since the Ocelot checks if it has the tempt goal, and this only happens on survival mode when a player is holding a fish.

### CODE REVIEW

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java [hunk 1]
```diff
@@ -130,4 +137,18 @@
...
+         * Invoked before an tamable animal is tamed.
+         * This event only works on vanilla mobs. If needed in a modded entity, the mod should register this event on tame.
+         * Equivalent to Forge's {@code AnimalTameEvent} event.
+         *
+         * @param player The tamer.
+         * @param animal The animal being tamed.
+         * @return A {@link InteractionResult} determining the outcome of the event,
+         * if an outcome is InteractionResult::FAIL, the animal isn't tamed.
+         */
+        InteractionResult onTame(Animal animal, Player player);
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Please use EventResult

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java [hunk 2]
```diff
@@ -130,4 +137,18 @@
...
         void enterSection(Entity entity, int sectionX, int sectionY, int sectionZ, int prevX, int prevY, int prevZ);
     }
+    
+    interface AnimalTame {
+        /**
+         * Invoked before an tamable animal is tamed.
+         * This event only works on vanilla mobs. If needed in a modded entity, the mod should register this event on tame.
+         * Equivalent to Forge's {@code AnimalTameEvent} event.
+         *
+         * @param player The tamer.
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: please order the parameters according to the actual parameters

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java [hunk 3]
```diff
@@ -130,4 +137,18 @@
          */
         void enterSection(Entity entity, int sectionX, int sectionY, int sectionZ, int prevX, int prevY, int prevZ);
     }
+    
+    interface AnimalTame {
+        /**
+         * Invoked before an tamable animal is tamed.
+         * This event only works on vanilla mobs. If needed in a modded entity, the mod should register this event on tame.
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]:
```suggestion
         * This event only works on vanilla mobs. Mods implementing their own entities may want to make their own events or invoke this.
```

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java [hunk 4]
```diff
@@ -53,6 +56,10 @@
      * @see EnterSection#enterSection(Entity, int, int, int, int, int, int)
      */
     Event<EnterSection> ENTER_SECTION = EventFactory.createLoop();
+    /**
+     * @see EnterSection#enterSection(Entity, int, int, int, int, int, int)
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Wrong method

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java [hunk 5]
```diff
@@ -130,4 +137,18 @@
          */
         void enterSection(Entity entity, int sectionX, int sectionY, int sectionZ, int prevX, int prevY, int prevZ);
     }
+    
+    interface AnimalTame {
+        /**
+         * Invoked before an tamable animal is tamed.
```
Comment (shedaniel) [inline common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: "a" or "an"? is "t" silent here?

**Review by shedaniel — CHANGES_REQUESTED** [review]

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java (hunk 5 above)
Comment (lazynessmind) [reply common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Done in 814fce1

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java (hunk 4 above)
Comment (lazynessmind) [reply common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Done in 814fce1

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java (hunk 3 above)
Comment (lazynessmind) [reply common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Done in 814fce1

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java (hunk 2 above)
Comment (lazynessmind) [reply common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Done in 814fce1

File: common/src/main/java/dev/architectury/event/events/common/EntityEvent.java (hunk 1 above)
Comment (lazynessmind) [reply common/src/main/java/dev/architectury/event/events/common/EntityEvent.java]: Done in 814fce1 an 3457d52

**Review by shedaniel — APPROVED** [review]

**Review by MaxNeedsSnacks — APPROVED** [review]

**Comment by MaxNeedsSnacks** [comment]
Side note: Should we backport this to 1.16 as well? Or should we generally only port 1.16 PRs to 1.17, but not the other way round

**Comment by shedaniel** [comment]
For a while I will backport when I merge the PR

**Comment by lazynessmind** [comment]
@shedaniel I tested on both versions and it works. I can port it to 1.16 and make a new pr

**Pull request merged**